lib/src/main/java/graphql/nadel/engine/NadelDeferSupport.kt

```diff
@@ -0,0 +1,167 @@
+package graphql.nadel.engine
+
+import graphql.incremental.DelayedIncrementalPartialResult
+import graphql.nadel.engine.NadelDeferSupport.OutstandingJobCounter.OutstandingJobHandle
+import graphql.nadel.engine.util.copy
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.launch
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+class NadelDeferSupport internal constructor(
+    private val delayedResultsChannel: Channel<DelayedIncrementalPartialResult> = Channel(UNLIMITED),
+) {
+    /**
+     * The root [Job] to actually run the defer work.
+     */
+    private val deferCoroutineJob = SupervisorJob()
+    private val deferCoroutineScope = CoroutineScope(deferCoroutineJob + Dispatchers.Default)
+
+    /**
+     * A single [Flow] that can only be collected from once.
+     */
+    private val resultFlow by lazy(delayedResultsChannel::consumeAsFlow)
+
+    init {
+        deferCoroutineJob.invokeOnCompletion {
+            require(outstandingJobCounter.isEmpty())
+            delayedResultsChannel.close()
+        }
+    }
+
+    private val outstandingJobCounter = OutstandingJobCounter()
+
+    fun defer(task: suspend CoroutineScope.() -> DelayedIncrementalPartialResult): Job {
+        return launch { outstandingJobHandle ->
+            val hasNext: Boolean
+            val result = try {
+                task()
+            } finally {
+                hasNext = outstandingJobHandle.decrementAndGetJobCount() > 0
+            }
+
+            delayedResultsChannel.send(
+                // Copy of result but with the correct hasNext according to the info we know
+                result.copy(hasNext = hasNext)
+            )
+        }
+    }
+
+    /**
+     * Assumes that the [Flow] will correctly set [DelayedIncrementalPartialResult.hasNext] correctly.
+     *
+     * i.e. that the last element in the [Flow] will have `hasNext` set to `false`.
+     */
+    fun defer(serviceResults: Flow<DelayedIncrementalPartialResult>): Job {
```

Comment:
Is this version of the `defer` function supposed to be used in the scenario where the underlying services return delayed responses? 
cc @sbarker2

Reply:
Yep we're using it in https://github.com/atlassian-labs/nadel/pull/540/files#diff-90d64a63cac8ddb0ce02278176c2fbc733244bdecab27b930281841f66d2eb7fR325-R329 now